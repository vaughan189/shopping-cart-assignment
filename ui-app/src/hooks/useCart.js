/* eslint-disable react-hooks/exhaustive-deps */
import { useEffect, useState } from "react";
import useAxios from "../api/useAxios";
import { ADD_TO_CART } from "../constants/endpoints";
import { cartService } from "../services";
import { formatCartData } from "../utils/cart";

const useCart = () => {
  const [cartProductList, setCartProductList] = useState([]);
  const {
    response: cartProductsData,
    loading: cartProductsDataLoader,
    fetchData: fetchCartProducts,
  } = useAxios();

  const {
    response: deleteCartProductData,
    // loading: deleteCartProductLoader,
    fetchData: deleteCartProduct,
  } = useAxios();

  const {
    response: addProductData,
    // loading: addProductDataLoader,
    fetchData: addProducts,
  } = useAxios();

  const getCartProducts = () => {
    fetchCartProducts({
      method: "GET",
      url: `${process.env.REACT_APP_SERVER_BASE_URL}${ADD_TO_CART}`,

      headers: {
        accept: "*/*",
      },
    });
  };

  const deleteCartItem = (id) => {
    deleteCartProduct({
      method: "DELETE",
      url: `${process.env.REACT_APP_SERVER_BASE_URL}${ADD_TO_CART}/${id}`,
      headers: {
        accept: "*/*",
      },
    });
  };

  const addProductsToCart = (item) => {
    addProducts({
      method: "POST",
      url: `${process.env.REACT_APP_SERVER_BASE_URL}${ADD_TO_CART}`,

      headers: {
        accept: "*/*",
      },
      data: {
        productId: item.productId || item.id,
        productName: item.name || item.productName,
        imageURL: item.imageURL,
        price: item.price,
        quantity: 1,
      },
    });
  };

  useEffect(() => {
    getCartProducts();
    cartService.sendCartData("Cart has been updated");
  }, [deleteCartProductData, addProductData]);

  useEffect(() => {
    if (
      cartProductsData &&
      cartProductsData.data &&
      cartProductsData.data.length > 0
    ) {
      setCartProductList(formatCartData(cartProductsData.data));
    } else setCartProductList([]);
  }, [cartProductsData, addProductData]);

  return {
    cartProductsData,
    cartProductList,
    cartProductsDataLoader,
    getCartProducts,
    addProductsToCart,
    deleteCartItem,
  };
};

export default useCart;
