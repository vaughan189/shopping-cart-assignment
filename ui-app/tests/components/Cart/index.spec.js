import React from "react";
import { shallow } from "enzyme";
import {
  Cart,
  CheckOutSection,
  CartHeader,
} from "../../../src/components/Cart";
import { withHooks } from "jest-react-hooks-shallow";

import useCart from "../../../src/hooks/useCart";
jest.mock("../../../src/hooks/useCart");

describe("Cart component", () => {
  it("should render the Cart component correctly when api is loading", () => {
    useCart.mockImplementation(() => {
      return {
        cartProductsData: [],
        cartProductList: [],
        cartProductsDataLoader: true,
        addProductsToCart: jest.fn(),
        deleteCartItem: jest.fn(),
      };
    });
    const component = shallow(<Cart debug />);
    expect(component).toMatchSnapshot();
  });

  it("should render the Cart component correctly when api returns data", () => {
    useCart.mockImplementation(() => {
      return {
        cartProductsData: [
          {
            productId: "5b6c6a7f01a7c38429530883",
            productName: "Fresho Kiwi - Green, 3 pcs",
            imageURL: "/static/images/products/fruit-n-veg/kiwi-green.jpg",
            price: 87,
            quantity: 1,
            id: 1,
          },
          {
            productId: "5b6c6a7f01a7c38429530883",
            productName: "Fresho Kiwi - Green, 3 pcs",
            imageURL: "/static/images/products/fruit-n-veg/kiwi-green.jpg",
            price: 87,
            quantity: 1,
            id: 1,
          },
          {
            productId: "5b6c6aeb01a7c38429530884",
            productName: "Apple - Washington, Regular, 4 pcs",
            imageURL: "/static/images/products/fruit-n-veg/apple.jpg",
            price: 187,
            quantity: 1,
            id: 2,
          },
        ],
        cartProductList: [
          {
            productId: "5b6c6a7f01a7c38429530883",
            productName: "Fresho Kiwi - Green, 3 pcs",
            imageURL: "/static/images/products/fruit-n-veg/kiwi-green.jpg",
            price: 87,
            quantity: 1,
            id: 1,
          },
          {
            productId: "5b6c6aeb01a7c38429530884",
            productName: "Apple - Washington, Regular, 4 pcs",
            imageURL: "/static/images/products/fruit-n-veg/apple.jpg",
            price: 187,
            quantity: 2,
            id: 2,
          },
        ],
        cartProductsDataLoader: false,
        addProductsToCart: jest.fn(),
        deleteCartItem: jest.fn(),
      };
    });
    const component = shallow(<Cart debug />);
    expect(component).toMatchSnapshot();
  });

  it("should render the Cart component correctly when api returns empty array", () => {
    useCart.mockImplementation(() => {
      return {
        cartProductsData: [],
        cartProductList: [],
        cartProductsDataLoader: false,
        addProductsToCart: jest.fn(),
        deleteCartItem: jest.fn(),
      };
    });
    const component = shallow(<Cart debug />);
    expect(component).toMatchSnapshot();
  });

  it("should render the checkout section correctly", () => {
    jest.spyOn(React, "useEffect").mockImplementation((f) => f());
    const cartProductsData = {
      data: [
        {
          productId: "5b6c6a7f01a7c38429530883",
          productName: "Fresho Kiwi - Green, 3 pcs",
          imageURL: "/static/images/products/fruit-n-veg/kiwi-green.jpg",
          price: 87,
          quantity: 1,
          id: 1,
        },
        {
          productId: "5b6c6a7f01a7c38429530883",
          productName: "Fresho Kiwi - Green, 3 pcs",
          imageURL: "/static/images/products/fruit-n-veg/kiwi-green.jpg",
          price: 87,
          quantity: 1,
          id: 1,
        },
      ],
    };
    withHooks(() => {
      const component = shallow(
        <CheckOutSection cartProductsData={cartProductsData} debug />
      );
      expect(component).toMatchSnapshot();
    });
  });

  it("should render the Cart Header component correctly", () => {
    const component = shallow(<CartHeader debug />);
    expect(component).toMatchSnapshot();
  });
});
