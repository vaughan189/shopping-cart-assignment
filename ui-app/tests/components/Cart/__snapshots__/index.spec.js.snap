// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component should render the Cart Header component correctly 1`] = `
<ForwardRef(Typography)
  component="div"
  sx={
    Object {
      "backgroundColor": Object {
        "md": "#000000",
        "sm": "#FFFFFF",
        "xl": "#000000",
        "xs": "#FFFFFF",
      },
      "color": Object {
        "md": "#FFFFFF",
        "sm": "#000000",
        "xl": "#FFFFFF",
        "xs": "#000000",
      },
      "fontFamily": "Dosis",
      "fontSize": 18,
      "fontWeight": "bold",
      "height": "40px",
      "paddingLeft": 2,
      "paddingTop": 2,
    }
  }
>
  My Cart
   
  <ForwardRef(Typography)
    component="span"
    sx={
      Object {
        "fontFamily": "Dosis",
        "fontSize": 16,
      }
    }
  >
    (1 item)
  </ForwardRef(Typography)>
</ForwardRef(Typography)>
`;

exports[`Cart component should render the Cart component correctly when api is loading 1`] = `
<ForwardRef(Box)
  sx={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "fontFamily": "Dosis",
    }
  }
>
  <ForwardRef(Box)
    sx={Object {}}
  >
    <CartHeader />
  </ForwardRef(Box)>
  <ForwardRef(Box)
    sx={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "fontFamily": "Dosis",
        "height": "100%",
        "justifyContent": "Centre",
      }
    }
  >
    <ForwardRef(Typography)
      sx={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "Dosis",
          "fontSize": 25,
          "fontWeight": 600,
          "justifyContent": "center",
        }
      }
    >
      No items in your cart
    </ForwardRef(Typography)>
    <ForwardRef(Typography)
      sx={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "Dosis",
          "fontSize": 18,
          "fontWeight": 400,
          "justifyContent": "center",
        }
      }
    >
      Your favourite items are just a click away
    </ForwardRef(Typography)>
    <ForwardRef(Button)
      sx={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "centre",
          "textDecoration": "none",
        }
      }
      variant="contained"
    >
      Start Shopping
    </ForwardRef(Button)>
  </ForwardRef(Box)>
</ForwardRef(Box)>
`;

exports[`Cart component should render the Cart component correctly when api returns data 1`] = `
<ForwardRef(Box)
  sx={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "fontFamily": "Dosis",
    }
  }
>
  <ForwardRef(Box)
    sx={Object {}}
  >
    <CartHeader />
  </ForwardRef(Box)>
  <ForwardRef(Box)
    sx={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "fontFamily": "Dosis",
        "height": "100%",
        "justifyContent": "Centre",
      }
    }
  >
    <ForwardRef(Typography)
      sx={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "Dosis",
          "fontSize": 25,
          "fontWeight": 600,
          "justifyContent": "center",
        }
      }
    >
      No items in your cart
    </ForwardRef(Typography)>
    <ForwardRef(Typography)
      sx={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "Dosis",
          "fontSize": 18,
          "fontWeight": 400,
          "justifyContent": "center",
        }
      }
    >
      Your favourite items are just a click away
    </ForwardRef(Typography)>
    <ForwardRef(Button)
      sx={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "centre",
          "textDecoration": "none",
        }
      }
      variant="contained"
    >
      Start Shopping
    </ForwardRef(Button)>
  </ForwardRef(Box)>
</ForwardRef(Box)>
`;

exports[`Cart component should render the Cart component correctly when api returns empty array 1`] = `
<ForwardRef(Box)
  sx={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "fontFamily": "Dosis",
    }
  }
>
  <ForwardRef(Box)
    sx={Object {}}
  >
    <CartHeader />
  </ForwardRef(Box)>
  <ForwardRef(Box)
    sx={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "fontFamily": "Dosis",
        "height": "100%",
        "justifyContent": "Centre",
      }
    }
  >
    <ForwardRef(Typography)
      sx={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "Dosis",
          "fontSize": 25,
          "fontWeight": 600,
          "justifyContent": "center",
        }
      }
    >
      No items in your cart
    </ForwardRef(Typography)>
    <ForwardRef(Typography)
      sx={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "fontFamily": "Dosis",
          "fontSize": 18,
          "fontWeight": 400,
          "justifyContent": "center",
        }
      }
    >
      Your favourite items are just a click away
    </ForwardRef(Typography)>
    <ForwardRef(Button)
      sx={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "centre",
          "textDecoration": "none",
        }
      }
      variant="contained"
    >
      Start Shopping
    </ForwardRef(Button)>
  </ForwardRef(Box)>
</ForwardRef(Box)>
`;

exports[`Cart component should render the checkout section correctly 1`] = `
<ForwardRef(Box)
  sx={
    Object {
      "alignItems": "centre",
      "backgroundColor": "white",
      "bottom": 0,
      "display": "flex",
      "flexDirection": "column",
      "justifyContent": "centre",
      "padding": 2,
    }
  }
>
  <ForwardRef(Typography)
    sx={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "fontFamily": "Dosis",
        "fontSize": 16,
        "fontWeight": 600,
        "justifyContent": "center",
      }
    }
  >
    Promo code can be applied on payment page
  </ForwardRef(Typography)>
  <ForwardRef(Button)
    sx={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
    variant="contained"
  >
    <ForwardRef(Typography)
      component="span"
    >
      Proceed to Checkout
    </ForwardRef(Typography)>
    <ForwardRef(Box)
      sx={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-around",
        }
      }
    >
      <ForwardRef(Typography)
        component="span"
        sx={
          Object {
            "textTransform": "none",
          }
        }
      >
        Rs.
      </ForwardRef(Typography)>
      <ForwardRef(IconButton)
        aria-label="delete"
        color="secondary"
        size="small"
      >
        <Memo(ForwardRef(ArrowForwardIosSharpIcon))
          fontSize="inherit"
          sx={
            Object {
              "color": "#FFFFFF",
            }
          }
        />
      </ForwardRef(IconButton)>
    </ForwardRef(Box)>
  </ForwardRef(Button)>
</ForwardRef(Box)>
`;
